libraries/eosiolib/contracts/eosio/key_value.hpp

```diff
@@ -0,0 +1,1018 @@
+            return ret_val;
+         }
+
+         void* buffer = value_size > detail::max_stack_buffer_size ? malloc(value_size) : alloca(value_size);
+         auto copy_size = internal_use_do_not_use::kv_get_data(config.db_name, 0, (char*)buffer, value_size);
+
+         void* deserialize_buffer = buffer;
+         size_t deserialize_size = copy_size;
+
+         if (config.index_name != config.primary_index_name) {
+            uint32_t actual_data_size;
+            auto success = internal_use_do_not_use::kv_get(config.db_name, config.contract_name.value, (char*)buffer, copy_size, actual_data_size);
+            eosio::check(success, "failure getting primary key");
+
+            void* pk_buffer = actual_data_size > detail::max_stack_buffer_size ? malloc(actual_data_size) : alloca(actual_data_size);
+            auto pk_copy_size = internal_use_do_not_use::kv_get_data(config.db_name, 0, (char*)pk_buffer, actual_data_size);
+
+            deserialize_buffer = pk_buffer;
+            deserialize_size = pk_copy_size;
+         }
+
+         ret_val.emplace();
+         deserialize(*ret_val, deserialize_buffer, deserialize_size);
+
+         if (value_size > detail::max_stack_buffer_size) {
+            free(buffer);
+         }
+
+         return ret_val;
+      }
+
+      /**
+       * Returns an iterator to the object with the lowest key (by this index) in the table.
+       * @ingroup keyvalue
+       *
+       * @return An iterator to the object with the lowest key (by this index) in the table.
+       */
+      iterator begin() {
+         uint32_t itr = internal_use_do_not_use::kv_it_create(config.db_name, config.contract_name.value, config.prefix.data(), config.prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_lower_bound(itr, "", 0);
+
+         return {itr, static_cast<kv_it_stat>(itr_stat), config};
+      }
+
+      /**
+       * Returns an iterator referring to the `past-the-end` element. It does not point to any element, therefore `value` should not be called on it.
+       * @ingroup keyvalue
+       *
+       * @return An iterator referring to the `past-the-end` element.
+       */
+      iterator end() {
+         return {0, kv_it_stat::iterator_end, config};
+      }
+
+      /**
+       * Returns an iterator pointing to the element with the lowest key greater than or equal to the given key.
+       * @ingroup keyvalue
+       *
+       * @return An iterator pointing to the element with the lowest key greater than or equal to the given key.
+       */
+      iterator lower_bound(const K& key) {
+         auto t_key = table_key(config.prefix, make_key(key));
+
+         uint32_t itr = internal_use_do_not_use::kv_it_create(config.db_name, config.contract_name.value, config.prefix.data(), config.prefix.size());
+         int32_t itr_stat = internal_use_do_not_use::kv_it_lower_bound(itr, t_key.data(), t_key.size());
+
+         if (static_cast<kv_it_stat>(itr_stat) == kv_it_stat::iterator_end) {
```

Comment:
Is this test useful?  The iterator is already at end.

Reply:
Yeah, looking at it now I don't think it actually adds any value.